### ISSUE TITLE
Just an idea: saving module groups?

### ISSUE THREAD

**Issue by martin-lueders** [VCVRack/Rack#209]
OS (if applicable): all
Version (or "dev" if compiling from source): all

Here is just an idea which could be useful for working with VCV rack, and should not be too hard to implement: I think it might be useful if one could save groups of modules including their internal wiring. Sometimes, we do use the same building blocks or combinations of modules in different patches.

What do you think?

**Comment by sanderbaan** [comment]
Sounds like a great idea, like Ableton Instrument Racks? with macro controls...

**Comment by antoniotuzzi** [comment]
a MAX subpatch ? with declared  IN and OUT ?

**Comment by martin-lueders** [comment]
@antoniotuzzi No, I was not thinking of putting them into a container, or something. Just a way to select a few modules of a patch and save them, with their internal cables, as a group.

I guess the only difficulty is, when loading such group, where to place all the modules, as the information stored in the json bit, will most likely conflict with existing modules in the patch, into which you want to insert such a group.

It would either require some sort of "intelligent placement" or simply dump them into the next free row.

**Comment by AndrewBelt** [comment]
>should not be too hard to implement
It would be a *nightmare* to implement, and even design.

**Comment by martin-lueders** [comment]
Well, what I suggested at first is _not_ a macro, but just the possibility to export and import (or save and load) groups of modules with their internal wires.

I agree that macros would be a huge task and a nightmare. What I had in mind would be much simpler.

I will have a look into the code to see what might be missing. Basically, saving and loading the group would be nearly the same as saving and loading whole patches, only that it affects only a subset of modules in the patch, and that it does not overwrite the current patch on load, but adds the new modules to the end.

The things missing are:

1) A mechanism to select modules
2) A way to insert them without conflict of positions. The easiest way would be to add them in a new, free row. at the bottom and then leave it to the user to move them where (s)he wants.

**Comment by AndrewBelt** [comment]
Maybe "Export rectangle" will be added to the File menu later. With Import, this would solve this issue.

**Comment by martin-lueders** [comment]
OK. This requires to move the modules to be selected into a rectangle (which is a good idea anyway). So, I think this should be ok.

**Comment by AndrewBelt** [comment]
No, selecting it would allow you to highlight a rectangle of modules, and then a Save dialog would be displayed.

**Comment by martin-lueders** [comment]
Oh, sorry, I was not clear. What I meant is that it requires that the modules to be selected fall into a rectangle. Meaning to say they can't be randomly placed in the rack.

That just requires a bit discipline in building patches, or rearranging them before exporting a group.

But it will work. :)

**Comment by LKHSogpit** [comment]
This would be very handy. Individual module templates!

**Comment by odahan** [comment]
Group concept can be very powerful. Macro seems to hard to implement. But group can be made "easily" just to : 1/ load/save groups, 2/ move groups as one object in the UI
Each module as at least 2 screws so we can use screws as group markers. It can be easy without any addition to the UI to give a color to screws of modules belonging to the same group. It will materialize the group without the need to create a special container for example.
Just an idea...

**Comment by gnapier3** [comment]
Just wanted to throw in my non-coder opinion in support of this idea. I think it would be incredibly helpful to be able to save a set of modules as a self-contained "patch" that could be loaded into an existing project/rack. Any connections that extend beyond the saved set of modules are simply "lost" and have to be re-established in the new rack upon instantiation. However any connections within the saved set of modules are preserved. Thanks.

**Comment by aptrn** [comment]
> 
> 
> Maybe "Export rectangle" will be added to the File menu later. With Import, this would solve this issue.

Actually it would be the feature set to really change the game for me. Importing and saving subsections of patches to abstract "usual tricks" into .vcv files and recall them while planning a more complex one. I've been dreaming about it for a while

**Comment by mattjoyce** [comment]
Perhaps one approach would be to provide a method of assigning a module to a group (right click, assign to group 1).   Then provide File\Export\Group, which would export those module and any connection between those modules.

**Comment by nkurence** [comment]
Came to request just this, a sort of macro-preset system where you can save out and import groups of modules including their state and wiring. Kind of an export / import function but more integrated and visual, ideally as part of the library (maybe under a 'Groups' tab) with (larger) previews.

I'm not sure what would be best for interaction. Marquee selection isn't so great as maybe the modules you want to group aren't adjacent so maybe a keyboard modifier (Shift) to allow module multiselection (indicated by a border or colored overlay of the module) with a special context menu for related commands ('Save Group' etc.) would work. (e.g. when multiselection, right-clicking a selected module shows a multiselection command menu)

In most cases I would want to organize the modules before exporting so they load back up in a nice way, and this may or may not have anything to do with the current rack I'm working on - so a duplicate function that deals with multiselection would be really helpful. e.g. where all the selected modules with their wiring intact get duplicated and placed next to each other in a row, where you could then rearrange them if desired before saving them as a group. It could be the standard duplicate command (easiest for me) or if you wanted to offer additional functions like 'Duplicate with Wiring', 'Duplicate without Wiring' etc. they could be new commands in the multiselection context menu. In this way you can create groups as you go without disturbing too much your current project organization

Seeing how Stoermelder Strip can see + interact with adjacent modules makes me think this could also somehow be worked into a kind of 'Storage Strip' module. Where the saving/loading of module groups is done by way of a macro module (e.g. place desired module group manually next to the macro module and save a module preset). From a user experience perspective this is nowhere near as nice; lacking visual reference, requiring additional steps, manual organization/duplication and re-patching, and usually many many more clicks - but it would be preferable to nothing, or maybe a useful trigger for another idea altogether.
